std/socket.d

```diff
@@ -289,56 +289,93 @@ bool wouldHaveBlocked() nothrow @nogc
+// Returns true iff getaddrinfo is supported on this platform.
+version(Windows) private @property bool getaddrinfoSupported()
 {
-    typeof(&getnameinfo) getnameinfoPointer;
-    typeof(&getaddrinfo) getaddrinfoPointer;
-    typeof(&freeaddrinfo) freeaddrinfoPointer;
+    initialize;
+    assert((getaddrinfoPointer is null) == (freeaddrinfoPointer is null));
+    return getaddrinfoPointer !is null;
 }
+// Ditto (always supported on Posix).
+version(Posix) private enum bool getaddrinfoSupported = true;
 
-shared static this() @system
+version(Windows) private
 {
-    version (Windows)
-    {
-        WSADATA wd;
+    import core.atomic;
 
-        // Winsock will still load if an older version is present.
-        // The version is just a request.
-        int val;
-        val = WSAStartup(0x2020, &wd);
-        if (val)         // Request Winsock 2.2 for IPv6.
-            throw new SocketOSException("Unable to initialize socket library", val);
+    shared typeof(&getnameinfo) getnameinfoPointer;
+    shared typeof(&getaddrinfo) getaddrinfoPointer;
+    shared typeof(&freeaddrinfo) freeaddrinfoPointer;
 
-        // These functions may not be present on older Windows versions.
-        // See the comment in InternetAddress.toHostNameString() for details.
-        auto ws2Lib = GetModuleHandleA("ws2_32.dll");
-        if (ws2Lib)
-        {
-            getnameinfoPointer = cast(typeof(getnameinfoPointer))
-                                 GetProcAddress(ws2Lib, "getnameinfo");
-            getaddrinfoPointer = cast(typeof(getaddrinfoPointer))
-                                 GetProcAddress(ws2Lib, "getaddrinfo");
-            freeaddrinfoPointer = cast(typeof(freeaddrinfoPointer))
-                                 GetProcAddress(ws2Lib, "freeaddrinfo");
-        }
+    auto our_getnameinfo(A...)(A a) @system
+    {
+        initialize;
+        return getnameinfoPointer(a);
     }
-    else version (Posix)
+
+    auto our_getaddrinfo(A...)(A a) @system
     {
-        getnameinfoPointer = &getnameinfo;
-        getaddrinfoPointer = &getaddrinfo;
-        freeaddrinfoPointer = &freeaddrinfo;
+        initialize;
+        return getaddrinfoPointer(a);
     }
-}
 
+    auto our_freeaddrinfo(A...)(A a) @system
+    {
+        initialize;
+        return freeaddrinfoPointer(a);
+    }
 
-shared static ~this() @system nothrow @nogc
-{
-    version (Windows)
+    void initialize() @trusted
     {
-        WSACleanup();
+        static shared bool processInitialized;
+        if (atomicLoad(processInitialized)) return;
+
+        import std.concurrency : initOnce;
+
+        static bool perform()
+        {
+            // Winsock will still load if an older version is present.
+            // The version is just a request.
+            WSADATA wd;
+            const val = WSAStartup(0x2020, &wd);
+            if (val)         // Request Winsock 2.2 for IPv6.
+                throw new SocketOSException("Unable to initialize socket library", val);
```

Comment:
I'm considering aborting the application here. It pretty much means the Windows installation is messed up. Thoughts?

Reply:
Sounds like an horrible idea for GUI applications. Why?

Reply:
There's just no realistic scenario that warrants throwing an exception (and forcing all transitive clients of this function to mind it). Last time `WSAStartup()` failed was on Windows 7 or whatever: https://www.youtube.com/watch?v=MWc6fGipU28.

Normally I wouldn't mind much, but it's time to acknowledge that throwing exceptions is painful for programmatic clients and we should only throw them when it makes sense. We need to tighten the screws on this stuff.

Reply:
Oh, there is an alternative: simply return and leave the pointers to functions to be null. Then nothing will work.

Reply:
OK, I think this is nice: if the lib doesn't load, no networking functionality is supported. There was code before that accounted for that anyway.